cmd/snapshot-controller/main.go

```diff
@@ -64,18 +64,17 @@ var (
 )
 
 // Checks that the VolumeSnapshot v1 CRDs exist.
-func ensureCustomResourceDefinitionsExist(kubeClient *kubernetes.Clientset, client *clientset.Clientset) error {
+func ensureCustomResourceDefinitionsExist(client *clientset.Clientset) error {
 	condition := func() (bool, error) {
 		var err error
-		_, err = kubeClient.CoreV1().Namespaces().Get(context.TODO(), "kube-system", metav1.GetOptions{})
-		if err == nil {
-			// only execute list VolumeSnapshots if the kube-system namespace exists
-			_, err = client.SnapshotV1().VolumeSnapshots("kube-system").List(context.TODO(), metav1.ListOptions{})
```

Comment:
I still think passing "" as the namespace should work: https://stackoverflow.com/questions/64191539/how-to-list-all-pods-in-k8s-cluster-using-client-go-in-golang-program

Reply:
I read the source code for `VolumeSnapshots("").List()` as an exercise and found out that the `namespace/{ns}` string is not set in the REST url if `namespace == ''`

```golang
func (c *SnapshotV1Client) VolumeSnapshots(namespace string) VolumeSnapshotInterface {
	return newVolumeSnapshots(c, namespace)
}

    func newVolumeSnapshots(c *SnapshotV1Client, namespace string) *volumeSnapshots {
	    return &volumeSnapshots{
	      	client: c.RESTClient(),
		    ns:     namespace,
	    }
    }

func (c *volumeSnapshots) List(ctx context.Context, opts metav1.ListOptions) (result *v1.VolumeSnapshotList, err error) {
	var timeout time.Duration
	if opts.TimeoutSeconds != nil {
		timeout = time.Duration(*opts.TimeoutSeconds) * time.Second
	}
	result = &v1.VolumeSnapshotList{}
	err = c.client.Get().
		Namespace(c.ns).
		Resource("volumesnapshots").
		VersionedParams(&opts, scheme.ParameterCodec).
		Timeout(timeout).
		Do(ctx).
		Into(result)
	return
}

    // Namespace applies the namespace scope to a request (<resource>/[ns/<namespace>/]<name>)
    func (r *Request) Namespace(namespace string) *Request {
	    if r.err != nil {
		    return r
    	}
	    if r.namespaceSet {
	    	r.err = fmt.Errorf("namespace already set to %q, cannot change to %q", r.namespace, namespace)
		    return r
	    }
    	if msgs := IsValidPathSegmentName(namespace); len(msgs) != 0 {
	    	r.err = fmt.Errorf("invalid namespace %q: %v", namespace, msgs)
		    return r
		}
		r.namespaceSet = true
		r.namespace = namespace
		return r
	}

	func (r *Request) Do(ctx context.Context) Result {
		var result Result
		err := r.request(ctx, func(req *http.Request, resp *http.Response) {
			result = r.transformResponse(resp, req)
		})
		if err != nil {
			return Result{err: err}
		}
		return result
	}

		func (r *Request) request(ctx context.Context, fn func(*http.Request, *http.Response)) error {
		    // ...
			for {
				url := r.URL().String()
				req, err := http.NewRequest(r.verb, url, r.body)

			// URL returns the current working URL.
			func (r *Request) URL() *url.URL {
				p := r.pathPrefix
				if r.namespaceSet && len(r.namespace) > 0 {
					p = path.Join(p, "namespaces", r.namespace)
				}

```

The part that skips it is `len(r.namespace) > 0`, I confirmed that the list works by using kubeproxy + curl

```
curl localhost:8001/apis/snapshot.storage.k8s.io/v1/volumesnapshots?limit=500
{"apiVersion":"snapshot.storage.k8s.io/v1","items":[],"kind":"VolumeSnapshotList","metadata":{"continue":"","resourceVersion":"13979098","selfLink":"/apis/snapshot.storage.k8s.io/v1/volumesnapshots"}}
```

I was wondering if `--all-namespaces` is doing the same thing i.e. if it's setting namespace to an empty string, thanks @msau42 for suggesting this!